### DISCUSSION TITLE
How does the bare-metal binary semaphore API work?

### DISCUSSION THREAD

**Discussion by cr1901** [hathach/tinyusb#1807] [Q&A]
_This is a minddump for future-me, @hathach. Sorry for the wall of text :P._

I can't remember why, but I ended up looking at some of parts of the TinyUSB stack I'm not familiar with. I can't convince myself that the [binary semaphore API](https://github.com/hathach/tinyusb/blob/159aa599be42534cd9a37ae03c71c228c3e72e90/src/osal/osal_none.h#L49-L76) works and I have a few questions:

1. Why does `osal_semaphore_post` take a `in_isr` argument that's ignored? Why does `osal_semaphore_wait` not have the same `in_isr` argument?
2. Mutexes in the bare-metal API say they are never used in an ISR context (though this may [not be true](https://github.com/hathach/tinyusb/blob/3ead682af05bee3a275125f3e0384dd9b2882546/src/osal/osal_pico.h#L141) for Pico)<sup>1</sup>. Is this also true for the binary semaphore bare-metal API?
3. Why are the binary semaphores initialized to `0` instead of `1` (while the bare-metal mutexes implemented in terms of the binary semaphores are initialized to `1`)?
4. Consider the following code in `osal_semaphore_wait`:
    ```
    while (sem_hdl->count == 0) { }
    sem_hdl->count--;
    ```

   What prevents two or more threads<sup>2</sup> from both seeing `sem_hdl->count` as `1` and all of them decrementing `count`?

And a bonus question :D:
5. Based on the above questions, could you "roll your own" timer peripheral-based context-switching (single or multiple CPUs) on top of `osal_none.h`, or are you expected to add a new `osal_foo.h` header? The fact that there's a `OSAL_MUTEX_REQUIRED` config even in `osal_none.h` makes me think "multiple tasks threads on bare metal with single or multiple CPUs" is possible. But the bare-metal mutex under that `OSAL_MUTEX_REQUIRED` config is implemented in terms of the binary semaphore API, and doesn't seem to handle the cases in footnote 2.

## Footnotes
1. In fact, mutexes [seem to be](https://github.com/hathach/tinyusb/blob/159aa599be42534cd9a37ae03c71c228c3e72e90/src/osal/osal_none.h#L105-L111) no-ops unless you have multiple cores.
2. Multiple threads could be, for instance:
    1. **ISR and one or more TinyUSB task threads**: An interrupt happens in between checking `count` and decrementing `count`, and the ISR runs. Then the ISR decrements `count` because [nothing can interrupt](https://github.com/hathach/tinyusb/blob/159aa599be42534cd9a37ae03c71c228c3e72e90/src/osal/osal_none.h#L167-L169) the ISR.
      
       On the other hand, if the task thread decrements `count` first, and an interrupt happens, then the interrupt handler will get stuck waiting for `count` to become nonzero forever. So maybe it's not a good idea to use a semaphore in an ISR :P.
    2. **Multiple TinyUSB task threads on a single CPU**: A context switch happens in a first task between checking `count` and decrementing `count` (see question 5). A second task then checks `count` and decrements `count` successfully, but doesn't call `osal_semaphore_post` before being forcefully being shifted out. The first then `decrements `count` successfully.
    3. **Multiple TinyUSB task threads on multiple CPUs _without atomics_**: If you have atomics, use those :P. If you don't, it's possible to use a [software mutual exclusion algorithm](https://en.wikipedia.org/wiki/Mutual_exclusion#Software_solutions), or hardware support (what RP2040 does), to create multicore-safe critical sections. However, without using one of the above, two tasks on different CPUs can read and then decrement `count` at the same time without each other knowing just through normal execution!
    
       Additionally, it is possible to use a [FIFO without any atomic ops except a load/store](https://en.wikipedia.org/wiki/Producer%E2%80%93consumer_problem#Without_semaphores_or_monitors) to synchronize data between _two_ cores, but this is probably not an appropriate synchronization primitive for the TinyUSB stack internals.

**Comment by hathach** [comment]
sorry for late response as usual, for the context:
- osal is internal code, and should only be used within tinyusb. Therefore all the osal API is only within the context of tinyusb usage.
- semaphore is kind of obsolete now, it is fully replaced by mutex, in the future semaphore API will probably be removed

> _This is a minddump for future-me, @hathach. Sorry for the wall of text :P._
> 
> I can't remember why, but I ended up looking at some of parts of the TinyUSB stack I'm not familiar with. I can't convince myself that the [binary semaphore API](https://github.com/hathach/tinyusb/blob/159aa599be42534cd9a37ae03c71c228c3e72e90/src/osal/osal_none.h#L49-L76) works and I have a few questions:

No problems at all, I sometimes found strange/weird code in the stack (written by me) as well :D

> 1. Why does `osal_semaphore_post` take a `in_isr` argument that's ignored? Why does `osal_semaphore_wait` not have the same `in_isr` argument?

in general, in no-RTOS, `in_isr` is not relevant in the stack usage scenario. Though semaphore is not used any more, it will be probably removed in the future.

> 2. Mutexes in the bare-metal API say they are never used in an ISR context (though this may [not be true](https://github.com/hathach/tinyusb/blob/3ead682af05bee3a275125f3e0384dd9b2882546/src/osal/osal_pico.h#L141) for Pico)1. Is this also true for the binary semaphore bare-metal API?

mutex more correctly osal_mutex_API() is never used in the ISR (note: ISR here means USB ISR). It does not cover the case where other RTOS such as OS_freeRTOS or in this case OS_PICOSDK implement an internal mutex for queue API. The note is added by RPI team when initially ported TinyUSB. I will revise and update this comment later. (it probably obsolete).   

> 3. Why are the binary semaphores initialized to `0` instead of `1` (while the bare-metal mutexes implemented in terms of the binary semaphores are initialized to `1`)?

It depends on the usage, previously in tinyusb stack, mutex always need to lock() then unlock() therefore initial count is 1. While semaphore is always post() first (by usb isr) then wait() by other task --> count = 0. In more generic usage of semaphore, initial count is normally user configurable when create an semaphore. 

> 4. Consider the following code in `osal_semaphore_wait`:
>    ```
>    while (sem_hdl->count == 0) { }
>    sem_hdl->count--;
>    ```
>    What prevents two or more threads2 from both seeing `sem_hdl->count` as `1` and all of them decrementing `count`?

Previously, I have assumed that without preempted RTOS, context is never switch within the while, except for USB ISR. The stack usage here is typically 1 main function wait for usb isr to call semaphore post(). And there is no `2 threads`. Of course, this assumption is obsolete with multiple cores MCU. That is why we are retiring semaphore API to use mutex. Most multiple cores MCU comes with IPC (interprocess communication) API from their vendor including mutex, we will favor those vendor/sdk function in the future. 
 
> And a bonus question :D: 5. Based on the above questions, could you "roll your own" timer peripheral-based context-switching (single or multiple CPUs) on top of `osal_none.h`, or are you expected to add a new `osal_foo.h` header? The fact that there's a `OSAL_MUTEX_REQUIRED` config even in `osal_none.h` makes me think "multiple tasks threads on bare metal with single or multiple CPUs" is possible. But the bare-metal mutex under that `OSAL_MUTEX_REQUIRED` config is implemented in terms of the binary semaphore API, and doesn't seem to handle the cases in footnote 2.

Short anwser no, osal should be a super thin layer just enough for the stack within in minimal usage:
- USB ISR -> event queue -> task
- Mutex to prevent endpoint and fifo conflict 

RTOS is complicated and timer requires additional hardware to run. USB actually has SOF which can use as timer, but I still prefer to skip it if I don't need to. `OSAL_MUTEX_REQUIRED` will use vendor sdk's IPC API in the future and then we could safely remove semaphore.

> ## Footnotes
> 1. In fact, mutexes [seem to be](https://github.com/hathach/tinyusb/blob/159aa599be42534cd9a37ae03c71c228c3e72e90/src/osal/osal_none.h#L105-L111) no-ops unless you have multiple cores.

Yes, single MCU without preempted RTOS has no race condition within tinyusb usage.

> 2. Multiple threads could be, for instance:

multiple threads (excluding usb isr) only possible with either multiple cores or an preempted RTOS. In osal none configuration, we will use IPC API in the future to address, if vendor sdk does not provide this (unlikely), we will probably use C atomic function

PS: I tried my best, but obviously couldn't answer it all. Feel free to correct me if I miss something with osal.

**Comment by cr1901** [comment]
Thank you for the thorough answers as always @hathach. I don't have any big questions now, but I'd like to keep the discussion open in case I think of more. It's probably a good thing once in a while to discuss internals to see where we are making implicit assumptions about how TinyUSB (is supposed to) works.

My questions were written as a minddump just before I want to sleep a few days ago. I knew then and there if I didn't write it out, I'd forget :D!

**Comment by hathach** [comment]
sure, I love these internal implementation discussion as well. Since some part of the code got obsolete here and there. This also give us a good chance to review/revise codes after a while.
